### ISSUE TITLE
apollo-kotlin fails when using refreshVersions

### ISSUE THREAD

**Issue by sproctor** [jmfayard/refreshVersions#507]
- [ :heavy_check_mark: ] I have read the [guidelines for submitting issues](https://jmfayard.github.io/refreshVersions/contributing/submitting-issues/)
## 🐛 Describe the bug
Apollo checks to make sure all versions are equal: https://github.com/apollographql/apollo-kotlin/blob/a1536a269ffe5896f1b5290e55bd361e2d398861/apollo-gradle-plugin/src/main/kotlin/com/apollographql/apollo3/gradle/internal/DefaultApolloExtension.kt#L167

This check fails when using refreshVersions.

This can be worked around by using `versionFor` in the dependencies.

## ⚠️ Current behavior
This results in the `generateApolloSources` task giving the following error:
```
Apollo: All apollo versions should be the same. Found:
  [3.1.0, _]
```

## ✅ Expected behavior
The version check to succeed.

## 💣 Steps to reproduce
Configuration:
```
buildscript {
    ...
    dependencies {
        ...
        classpath("com.apollographql.apollo3:apollo-gradle-plugin:_")
    }
}
```

```
plugins {
    ...
    id("com.apollographql.apollo3")
}

dependencies {
    implementation("com.apollographql.apollo3:apollo-runtime:_")
    implementation("com.apollographql.apollo3:apollo-adapters:_")
}
```

**Comment by LouisCAD** [comment]
Hello,
Can you show what are the related keys you have in the `versions.properties` file?

Regardless, I think we need to add rules for apollo-kotlin to synchronize the version of the plugin and of the dependencies, as we're doing for Kotlin and the Android Gradle plugin.

Maybe it's also time to work on a generalization of this version key sharing for plugin + dependencies instead of hardcoding this special case.

**Comment by sproctor** [comment]
Currently they are:
```
version.com.apollographql.apollo3..apollo-gradle-plugin=3.1.0

version.com.apollographql.apollo3..apollo-runtime=version.com.apollographql.apollo3..apollo-gradle-plugin
version.com.apollographql.apollo3..apollo-adapters=version.com.apollographql.apollo3..apollo-gradle-plugin
```

I've also tried:
```
version.com.apollographql.apollo3..apollo-gradle-plugin=3.1.0

version.com.apollographql.apollo3..apollo-runtime=3.1.0
version.com.apollographql.apollo3..apollo-adapters=3.1.0
```

**Comment by LouisCAD** [comment]
The way it works, it looks like refreshVersions can't do anything about it because of non-deterministic timing.

@martinbonnin Do you think otherwise? If not, I think it'd be great to have apollo-kotlin handle the version placeholder case, or provide a way to disable this check (and possibly disable it when refreshVersions is there). I'm open to discuss about the issue and the possibilities if it indeed implies changes in apollo-kotlin.

**Comment by LouisCAD** [comment]
Accessing `resolvedConfiguration` on a Gradle `Configuration` object would give you access to the information after everything happened, but I'll quote the side effects of running this from the Javadoc of `resolvedConfiguration`:

> Resolves this configuration. This locates and downloads the files which make up this configuration, and returns a `ResolvedConfiguration` that may be used to determine information about the resolve (including errors)

It might also be possible to use `resolutionStrategy` if it is triggered after `withDependencies. Those are all symbols of Gradle's `Configuration` type.

**Comment by martinbonnin** [comment]
Fix is there: https://github.com/apollographql/apollo-kotlin/pull/3896

As always with Gradle lifecycle, I'm crossing fingers this doesn't trigger side effects in other plugins that required the resolution to be done later

```kotlin
tasks.configureEach {
  if (name == "checkApolloVersions") {
    enabled = false
  }
}
```

**Comment by martinbonnin** [comment]
Thinking more about this, I'm having second thoughts... Resolving a configuration isn't something trivial and I'm too afraid of potential side effects. Also, if for some reason, another unrelated dependency fails to resolve, this will show up with "checkApolloVersions" in the error, which could be misleading.

I'm ok disabling the check if `de.fayard.refreshVersions` is applied (or maybe just filtering out '_'). What do you think?

**Comment by imashnake0** [comment]
Fixed in [3.2.0](https://github.com/apollographql/apollo-kotlin/releases/tag/v3.2.0) 🎉.
